sputnikdao2/src/upgrade.rs

```diff
@@ -0,0 +1,134 @@
+//! Logic to upgrade Sputnik contracts.
+
+use near_sdk::Gas;
+
+use crate::*;
+
+const FACTORY_KEY: &[u8; 7] = b"FACTORY";
+const DEFAULT_FACTORY_ID: &str = "sputnik-dao.near";
+const ERR_MUST_BE_SELF_OR_FACTORY: &str = "ERR_MUST_BE_SELF_OR_FACTORY";
+const SELF_MIGRATE_METHOD_NAME: &[u8; 7] = b"migrate";
+const UPDATE_GAS_LEFTOVER: Gas = Gas(5_000_000_000_000);
+const NO_DEPOSIT: Balance = 0;
+
+/// Gas for upgrading this contract on promise creation + deploying new contract.
+pub const GAS_FOR_UPGRADE_SELF_DEPLOY: Gas = Gas(30_000_000_000_000);
+
+pub const GAS_FOR_UPGRADE_REMOTE_DEPLOY: Gas = Gas(10_000_000_000_000);
+
+/// Info about factory that deployed this contract and if auto-update is allowed.
+#[derive(BorshSerialize, BorshDeserialize, Serialize, Deserialize)]
+#[cfg_attr(not(target_arch = "wasm32"), derive(Clone, Debug))]
+#[serde(crate = "near_sdk::serde")]
+pub struct FactoryInfo {
+    pub factory_id: AccountId,
+    pub auto_update: bool,
+}
+
+/// Fetches factory info from the storage.
+/// By design not using contract STATE to allow for upgrade of stuck contracts from factory.
+pub(crate) fn internal_get_factory_info() -> FactoryInfo {
+    env::storage_read(FACTORY_KEY)
+        .map(|value| FactoryInfo::try_from_slice(&value).expect("INTERNAL_FAIL"))
+        .unwrap_or_else(|| FactoryInfo {
+            factory_id: AccountId::new_unchecked(DEFAULT_FACTORY_ID.to_string()),
+            auto_update: true,
+        })
+}
+
+pub(crate) fn internal_set_factory_info(factory_info: &FactoryInfo) {
+    env::storage_write(
+        FACTORY_KEY,
+        &factory_info.try_to_vec().expect("INTERNAL_FAIL"),
+    );
+}
+
+/// Function that receives new contract, updates and calls migration.
+/// Two options who call it:
+///  - current account, in case of fetching contract code from factory;
+///  - factory, if this contract allows to factory-update;
+#[no_mangle]
+pub fn update() {
+    env::setup_panic_hook();
+    let factory_info = internal_get_factory_info();
+    let current_id = env::current_account_id();
+    assert!(
+        env::predecessor_account_id() == current_id
+            || (env::predecessor_account_id() == factory_info.factory_id
+                && factory_info.auto_update),
+        "{}",
+        ERR_MUST_BE_SELF_OR_FACTORY
+    );
+    unsafe {
+        // Load code into register 0 result from the promise.
+        sys::promise_result(0, 0);
+        // Update current contract with code from register 0.
+        let promise_id = sys::promise_batch_create(
+            current_id.as_bytes().len() as _,
+            current_id.as_bytes().as_ptr() as _,
+        );
+        // Deploy the contract code.
+        sys::promise_batch_action_deploy_contract(promise_id, u64::MAX as _, 0);
+        // Call promise to migrate the state.
+        // Batched together to fail upgrade if migration fails.
+        sys::promise_batch_action_function_call(
```

Comment:
I probably don't get this right, but it's not crystal clear to me why the `migrate` call is necessary here, since `migrate` will only read the contract code?

Is this to make sure that the `deploy` call succeeds? If that's the case, the `deploy` call can fail but the `migrate` call can succeed. Ex: if the contract code was already deployed and you're trying to deploy a newer version that fails, the `migrate` will succeed and return the old code.

Wouldn't be better to use `promise_batch_then` instead of `promise_batch_action_function_call` to make sure that deploy actually succeeded?

Reply:
current `migrate` is just a template for the actual migrate.
I've updated it to just be a better no-op.

Generally, when new version of the contract will be done - migrate should load the old state and save new state.

In case of migration function failing for whatever reason, if `promise_batch_then` was used - the contract will stay with the new contract code and old state. With single batch of two actions - if the second action fails, it will revert the first action (deploy of the contract) - recovering back to pre-upgrade state.